crypto/ocsp/ocsp_http.c

```diff
@@ -57,3 +61,64 @@ void OCSP_REQ_CTX_free(OCSP_REQ_CTX *rctx)
     OPENSSL_free(rctx->iobuf);
     OPENSSL_free(rctx);
 }
+
+OCSP_REQ_CTX *OCSP_sendreq_new(BIO *io, const char *path, OCSP_REQUEST *req,
+                               int maxline)
+{
+
+    OCSP_REQ_CTX *rctx = NULL;
+    rctx = OCSP_REQ_CTX_new(io, maxline);
+    if (rctx == NULL) {
+      return NULL;
+    }
+
+    if (!OCSP_REQ_CTX_http(rctx, "POST", path)) {
+      goto err;
+    }
+    if (req != NULL && !OCSP_REQ_CTX_set1_req(rctx, req)) {
+      goto err;
+    }
+    return rctx;
+
+ err:
+    OCSP_REQ_CTX_free(rctx);
+    return NULL;
+}
+
+int OCSP_REQ_CTX_http(OCSP_REQ_CTX *rctx, const char *op, const char *path)
+{
+    static const char http_hdr[] = "%s %s HTTP/1.0\r\n";
```

Comment:
This is a windows line ending, will this impact linux clients/servers?

Reply:
The `'\r\n` is part of the HTTP standard, this looks like this is writing out the initial request line. https://www.rfc-editor.org/rfc/rfc2616#section-5.1

Reply:
I still question why we need to be doing anything with HTTP in our library though? Shouldn't it be enough to provide serialization / de-serialization of the ASN1 messages?

There are two layers of OCSP:
1. There is the OCSP "protocol messages" which are request/response structure messages defined using ASN.1 syntax that are DER-encoded.
2. There are transports (for example HTTP) for which these protocol messages can be sent over. When using those transports there is defined behavior of how the message should be sent. (GET vs POST behavior etc). The transport mechanism for HTTP is defined in the Appendix of the RFC, but as the RFC states technically any transport could be used. [OCSP over HTTP](https://datatracker.ietf.org/doc/html/rfc6960#page-30)

I certainly think part 1 is the most difficult item to get right, and the one we should be providing APIs for. I question why we need to be doing item number two though. It seems crazy to me we are also synthesizing the raw http request bytes. Feels like that should be up to the application higher up in the stack to handle? Number two isn't what is difficult to implement OCSP for, the difficulty is in bullet point number one. It's a weird surface area to be adding to our library.

Reply:
Yeah I completely agree. I was hoping we could be focusing on part 1 for AWS-LC's OCSP implementation, and it initially was during our previous iterations of OCSP. Sadly though, OpenSSL implements the full package of OCSP inside their libcrypto layer, and teams moving over from OpenSSL to AWS-LC have continuously expected us to support OCSP for their use-case as well.
A good number of potential downstream consumers, like nginx or haproxy, implement their own HTTP behavior around OpenSSL's OCSP request APIs. But there are also a great deal of customers asks for us to support the OpenSSL1.1.1's original OCSP/HTTP behavior because they've been relying on it. A couple examples of customers that are already asking for the support right now are s2n-tls and the Datasync team. There are also findings of 7-8 internal packages relying on this behavior as well, which I suspect we're going to need to support as AWS-LC gains more adoption. It's a weird surface area to add, but this has been a part of a continuously growing ask from customers asking for parity with OpenSSL1.1.1.